File: zig/conformance_test.zig
/// SKG conformance tests — validates the Zig parser against shared testdata/ fixtures.
///
/// Each valid fixture is a .skg file with a .expected.json file describing the expected AST.
/// Each invalid fixture is a .skg file that should fail to parse, with a .expected.json
/// describing the expected error message substring.
const std = @import("std");
const testing = std.testing;
const skg_root = @import("root.zig");
const ast = @import("ast.zig");

// ─── Valid fixtures ─────────────────────────────────────────────────────────

const valid_fixtures = [_][]const u8{
    "all-types",
    "arrays",
    "block",
    "comments",
    "duplicate-lastwins",
    "empty",
    "escaped-string",
    "imports",
    "multiline-string",
    "nested-array",
    "nested-block",
    "simple-string",
    "versions",
};

const invalid_fixtures = [_][]const u8{
    "bad-escape",
    "duplicate-skg-version",
    "missing-colon",
    "mixed-array",
    "unclosed-block",
    "unterminated-array",
    "unterminated-string",
};

fn readFixture(alloc: std.mem.Allocator, comptime subdir: []const u8, name: []const u8, comptime ext: []const u8) ![]u8 {
    var path_buf: [512]u8 = undefined;
    const path = std.fmt.bufPrint(&path_buf, "testdata/{s}/{s}{s}", .{ subdir, name, ext }) catch return error.PathTooLong;

    const cwd = std.fs.cwd();
    const file = cwd.openFile(path, .{}) catch |err| {
        std.debug.print("Failed to open {s}: {}\n", .{ path, err });
        return err;
    };
    defer file.close();
    return file.readToEndAlloc(alloc, 1024 * 1024);
}

// ─── JSON parsing helpers ───────────────────────────────────────────────────

fn expectJsonString(json: std.json.Value) ?[]const u8 {
    return switch (json) {
        .string => |s| s,
        else => null,
    };
}

fn expectJsonBool(json: std.json.Value) ?bool {
    return switch (json) {
        .bool => |b| b,
        else => null,
    };
}

fn expectJsonInt(json: std.json.Value) ?i64 {
    return switch (json) {
        .integer => |n| n,
        else => null,
    };
}

fn expectJsonFloat(json: std.json.Value) ?f64 {
    return switch (json) {
        .float => |f| f,
        .integer => |n| @as(f64, @floatFromInt(n)),
        else => null,
    };
}

fn expectJsonArray(json: std.json.Value) ?[]std.json.Value {
    return switch (json) {
        .array => |a| a.items,
        else => null,
    };
}

fn expectJsonObject(json: std.json.Value) ?std.json.ObjectMap {
    return switch (json) {
        .object => |o| o,
        else => null,
    };
}

// ─── AST comparison ─────────────────────────────────────────────────────────

fn compareValue(expected_obj: std.json.ObjectMap, actual: ast.Value) !void {
    const type_str = expectJsonString(expected_obj.get("type") orelse return error.MissingType) orelse return error.BadType;

    if (std.mem.eql(u8, type_str, "string")) {
        try testing.expectEqual(ast.ValueType.string, std.meta.activeTag(actual));
        const expected_data = expectJsonString(expected_obj.get("data") orelse return error.MissingData) orelse return error.BadData;
        try testing.expectEqualStrings(expected_data, actual.string);
    } else if (std.mem.eql(u8, type_str, "int")) {
        try testing.expectEqual(ast.ValueType.int, std.meta.activeTag(actual));
        const expected_data = expectJsonInt(expected_obj.get("data") orelse return error.MissingData) orelse return error.BadData;
        try testing.expectEqual(expected_data, actual.int);
    } else if (std.mem.eql(u8, type_str, "float")) {
        try testing.expectEqual(ast.ValueType.float, std.meta.activeTag(actual));
        const expected_data = expectJsonFloat(expected_obj.get("data") orelse return error.MissingData) orelse return error.BadData;
        try testing.expectApproxEqAbs(expected_data, actual.float, 1e-9);
    } else if (std.mem.eql(u8, type_str, "bool")) {
        try testing.expectEqual(ast.ValueType.bool, std.meta.activeTag(actual));
        const expected_data = expectJsonBool(expected_obj.get("data") orelse return error.MissingData) orelse return error.BadData;
        try testing.expectEqual(expected_data, actual.bool);
    } else if (std.mem.eql(u8, type_str, "null")) {
        try testing.expectEqual(ast.ValueType.@"null", std.meta.activeTag(actual));
    } else if (std.mem.eql(u8, type_str, "array")) {
        try testing.expectEqual(ast.ValueType.array, std.meta.activeTag(actual));
        const expected_items = expectJsonArray(expected_obj.get("data") orelse return error.MissingData) orelse return error.BadData;
        try testing.expectEqual(expected_items.len, actual.array.items.len);
        for (expected_items, 0..) |item_json, i| {
            const item_obj = expectJsonObject(item_json) orelse return error.BadArrayItem;
            try compareValue(item_obj, actual.array.items[i]);
        }
    } else {
        std.debug.print("Unknown type in expected JSON: {s}\n", .{type_str});
        return error.UnknownType;
    }
}

fn compareNodes(expected_children: []std.json.Value, actual_children: []const ast.Node) !void {
    try testing.expectEqual(expected_children.len, actual_children.len);

    for (expected_children, 0..) |child_json, i| {
        const child_obj = expectJsonObject(child_json) orelse return error.BadChild;
        const node_type = expectJsonString(child_obj.get("type") orelse return error.MissingNodeType) orelse return error.BadNodeType;

        if (std.mem.eql(u8, node_type, "field")) {
            const actual = actual_children[i];
            try testing.expect(actual == .field);
            const expected_key = expectJsonString(child_obj.get("key") orelse return error.MissingKey) orelse return error.BadKey;
            try testing.expectEqualStrings(expected_key, actual.field.key);

            if (child_obj.get("value")) |val_json| {
                const val_obj = expectJsonObject(val_json) orelse return error.BadValue;
                try compareValue(val_obj, actual.field.value);
            }
        } else if (std.mem.eql(u8, node_type, "block")) {
            const actual = actual_children[i];
            try testing.expect(actual == .block);
            const expected_name = expectJsonString(child_obj.get("name") orelse return error.MissingName) orelse return error.BadName;
            try testing.expectEqualStrings(expected_name, actual.block.name);

            if (child_obj.get("children")) |children_json| {
                const nested = expectJsonArray(children_json) orelse return error.BadChildren;
                try compareNodes(nested, actual.block.children);
            }
        }
    }
}

fn compareNullableString(expected_json: std.json.Value, actual: ?[]const u8) !void {
    switch (expected_json) {
        .null => try testing.expect(actual == null),
        .string => |s| {
            try testing.expect(actual != null);
            try testing.expectEqualStrings(s, actual.?);
        },
        else => return error.BadExpectedJson,
    }
}

// ─── Test runners ───────────────────────────────────────────────────────────

fn runValidFixture(name: []const u8) !void {
    const skg_data = try readFixture(testing.allocator, "valid", name, ".skg");
    defer testing.allocator.free(skg_data);

    const json_data = try readFixture(testing.allocator, "valid", name, ".expected.json");
    defer testing.allocator.free(json_data);

    const parsed_json = std.json.parseFromSlice(std.json.Value, testing.allocator, json_data, .{}) catch |err| {
        std.debug.print("FAIL valid/{s}: bad JSON: {}\n", .{ name, err });
        return error.BadFixtureJson;
    };
    defer parsed_json.deinit();

    const root_obj = expectJsonObject(parsed_json.value) orelse return error.BadRoot;

    var result = skg_root.parseSource(testing.allocator, skg_data, name);
    defer result.deinit();

    if (result.file == null) {
        std.debug.print("FAIL valid/{s}: parse failed", .{name});
        if (result.diagnostic) |d| {
            std.debug.print(": {s}:{d}:{d}: {s}", .{ d.path, d.line, d.col, d.message });
        }
        std.debug.print("\n", .{});
        return error.ParseFailed;
    }
    const file = result.file.?;

    // Compare skg_version
    if (root_obj.get("skg_version")) |v| {
        try compareNullableString(v, file.skg_version);
    }

    // Compare schema_version
    if (root_obj.get("schema_version")) |v| {
        try compareNullableString(v, file.schema_version);
    }

    // Compare imports
    if (root_obj.get("imports")) |imports_json| {
        if (expectJsonArray(imports_json)) |expected_imports| {
            try testing.expectEqual(expected_imports.len, file.import_paths.len);
            for (expected_imports, 0..) |imp_json, i| {
                const imp_str = expectJsonString(imp_json) orelse return error.BadImport;
                try testing.expectEqualStrings(imp_str, file.import_paths[i]);
            }
        }
    }

    // Compare children
    if (root_obj.get("children")) |children_json| {
        const expected_children = expectJsonArray(children_json) orelse return error.BadChildren;
        compareNodes(expected_children, file.children) catch |err| {
            std.debug.print("FAIL valid/{s}: children mismatch: {}\n", .{ name, err });
            return err;
        };
    }
}

test "conformance: valid fixtures" {
    for (valid_fixtures) |name| {
        runValidFixture(name) catch |err| {
            std.debug.print("FAIL valid/{s}: {}\n", .{ name, err });
            return err;
        };
    }
}

fn runInvalidFixture(name: []const u8) !void {
    const skg_data = try readFixture(testing.allocator, "invalid", name, ".skg");
    defer testing.allocator.free(skg_data);

    const json_data = try readFixture(testing.allocator, "invalid", name, ".expected.json");
    defer testing.allocator.free(json_data);

    const parsed_json = std.json.parseFromSlice(std.json.Value, testing.allocator, json_data, .{}) catch |err| {
        std.debug.print("FAIL invalid/{s}: bad JSON: {}\n", .{ name, err });
        return error.BadFixtureJson;
    };
    defer parsed_json.deinit();

    const root_obj = expectJsonObject(parsed_json.value) orelse return error.BadRoot;

    var result = skg_root.parseSource(testing.allocator, skg_data, name);
    defer result.deinit();

    // Should have failed
    if (result.file != null) {
        std.debug.print("FAIL invalid/{s}: expected parse error, got success\n", .{name});
        return error.ExpectedParseError;
    }

    // Check message_contains if specified
    if (root_obj.get("message_contains")) |mc_json| {
        const needle = expectJsonString(mc_json) orelse return error.BadMessageContains;
        const diag = result.diagnostic orelse {
            std.debug.print("FAIL invalid/{s}: no diagnostic\n", .{name});
            return error.NoDiagnostic;
        };
        if (std.mem.indexOf(u8, diag.message, needle) == null) {
            std.debug.print("FAIL invalid/{s}: message \"{s}\" does not contain \"{s}\"\n", .{ name, diag.message, needle });
            return error.MessageMismatch;
        }
    }
}

test "conformance: invalid fixtures" {
    for (invalid_fixtures) |name| {
        runInvalidFixture(name) catch |err| {
            std.debug.print("FAIL invalid/{s}: {}\n", .{ name, err });
            return err;
        };
    }
}

File: go/conformance_test.go
package skg

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Conformance tests: validate Go parser against shared testdata/ fixtures.

type expectedFile struct {
	SKGVersion    *string        `json:"skg_version"`
	SchemaVersion *string        `json:"schema_version"`
	Imports       []string       `json:"imports"`
	Children      []expectedNode `json:"children"`
}

type expectedNode struct {
	Type     string         `json:"type"`     // "field" or "block"
	Key      string         `json:"key"`      // field key
	Name     string         `json:"name"`     // block name
	Value    *expectedValue `json:"value"`    // field value
	Children []expectedNode `json:"children"` // block children
}

type expectedValue struct {
	Type        string          `json:"type"` // "string", "int", "float", "bool", "null", "array"
	Data        json.RawMessage `json:"data"`
	ElementType string          `json:"element_type"`
}

type expectedError struct {
	Error           bool   `json:"error"`
	MessageContains string `json:"message_contains"`
}

func testdataDir() string {
	// testdata/ is one level up from go/
	return filepath.Join("..", "testdata")
}

func TestConformanceValid(t *testing.T) {
	dir := filepath.Join(testdataDir(), "valid")
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("cannot read valid fixtures dir: %v", err)
	}

	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".skg") {
			continue
		}
		name := strings.TrimSuffix(e.Name(), ".skg")
		t.Run(name, func(t *testing.T) {
			skgPath := filepath.Join(dir, e.Name())
			jsonPath := filepath.Join(dir, name+".expected.json")

			skgData, err := os.ReadFile(skgPath)
			if err != nil {
				t.Fatal(err)
			}
			jsonData, err := os.ReadFile(jsonPath)
			if err != nil {
				t.Fatal(err)
			}

			var expected expectedFile
			if err := json.Unmarshal(jsonData, &expected); err != nil {
				t.Fatalf("bad expected JSON: %v", err)
			}

			file, err := Parse(skgData)
			if err != nil {
				t.Fatalf("parse failed: %v", err)
			}

			// Compare skg_version
			if expected.SKGVersion == nil {
				if file.SKGVersion != nil {
					t.Errorf("expected skg_version nil, got %q", *file.SKGVersion)
				}
			} else {
				if file.SKGVersion == nil || *file.SKGVersion != *expected.SKGVersion {
					t.Errorf("skg_version: expected %q, got %v", *expected.SKGVersion, file.SKGVersion)
				}
			}

			// Compare schema_version
			if expected.SchemaVersion == nil {
				if file.SchemaVersion != nil {
					t.Errorf("expected schema_version nil, got %q", *file.SchemaVersion)
				}
			} else {
				if file.SchemaVersion == nil || *file.SchemaVersion != *expected.SchemaVersion {
					t.Errorf("schema_version: expected %q, got %v", *expected.SchemaVersion, file.SchemaVersion)
				}
			}

			// Compare imports
			if len(expected.Imports) != len(file.ImportPaths) {
				t.Errorf("imports: expected %d, got %d", len(expected.Imports), len(file.ImportPaths))
			} else {
				for i, imp := range expected.Imports {
					if file.ImportPaths[i] != imp {
						t.Errorf("import[%d]: expected %q, got %q", i, imp, file.ImportPaths[i])
					}
				}
			}

			// Compare children
			compareNodes(t, "", expected.Children, file.Children)
		})
	}
}

func TestConformanceInvalid(t *testing.T) {
	dir := filepath.Join(testdataDir(), "invalid")
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("cannot read invalid fixtures dir: %v", err)
	}

	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".skg") {
			continue
		}
		name := strings.TrimSuffix(e.Name(), ".skg")
		t.Run(name, func(t *testing.T) {
			skgPath := filepath.Join(dir, e.Name())
			jsonPath := filepath.Join(dir, name+".expected.json")

			skgData, err := os.ReadFile(skgPath)
			if err != nil {
				t.Fatal(err)
			}
			jsonData, err := os.ReadFile(jsonPath)
			if err != nil {
				t.Fatal(err)
			}

			var expected expectedError
			if err := json.Unmarshal(jsonData, &expected); err != nil {
				t.Fatalf("bad expected JSON: %v", err)
			}

			_, parseErr := Parse(skgData)
			if parseErr == nil {
				t.Fatal("expected parse error, got success")
			}

			if expected.MessageContains != "" {
				if !strings.Contains(parseErr.Error(), expected.MessageContains) {
					t.Errorf("error %q does not contain %q", parseErr.Error(), expected.MessageContains)
				}
			}
		})
	}
}

func compareNodes(t *testing.T, path string, expected []expectedNode, actual []Node) {
	t.Helper()
	if len(expected) != len(actual) {
		t.Errorf("%schildren count: expected %d, got %d", path, len(expected), len(actual))
		return
	}
	for i, en := range expected {
		an := actual[i]
		prefix := fmt.Sprintf("%s[%d].", path, i)

		switch en.Type {
		case "field":
			if an.Field == nil {
				t.Errorf("%sexpected field, got block", prefix)
				continue
			}
			if an.Field.Key != en.Key {
				t.Errorf("%skey: expected %q, got %q", prefix, en.Key, an.Field.Key)
			}
			if en.Value != nil {
				compareValue(t, prefix+"value.", *en.Value, an.Field.Value)
			}

		case "block":
			if an.Block == nil {
				t.Errorf("%sexpected block, got field", prefix)
				continue
			}
			if an.Block.Name != en.Name {
				t.Errorf("%sname: expected %q, got %q", prefix, en.Name, an.Block.Name)
			}
			compareNodes(t, prefix, en.Children, an.Block.Children)
		}
	}
}

func compareValue(t *testing.T, path string, expected expectedValue, actual Value) {
	t.Helper()

	// Map expected type string to ValueType
	var expectedType ValueType
	switch expected.Type {
	case "string":
		expectedType = TypeString
	case "int":
		expectedType = TypeInt
	case "float":
		expectedType = TypeFloat
	case "bool":
		expectedType = TypeBool
	case "null":
		expectedType = TypeNull
	case "array":
		expectedType = TypeArray
	default:
		t.Errorf("%sunknown expected type %q", path, expected.Type)
		return
	}

	if actual.Type != expectedType {
		t.Errorf("%stype: expected %v, got %v", path, expectedType, actual.Type)
		return
	}

	switch expected.Type {
	case "string":
		var s string
		if err := json.Unmarshal(expected.Data, &s); err != nil {
			t.Errorf("%scannot parse expected string data: %v", path, err)
			return
		}
		if actual.Str != s {
			t.Errorf("%svalue: expected %q, got %q", path, s, actual.Str)
		}

	case "int":
		var n float64 // JSON numbers are float64
		if err := json.Unmarshal(expected.Data, &n); err != nil {
			t.Errorf("%scannot parse expected int data: %v", path, err)
			return
		}
		if actual.Int != int64(n) {
			t.Errorf("%svalue: expected %d, got %d", path, int64(n), actual.Int)
		}

	case "float":
		var f float64
		if err := json.Unmarshal(expected.Data, &f); err != nil {
			t.Errorf("%scannot parse expected float data: %v", path, err)
			return
		}
		if math.Abs(actual.Float-f) > 1e-9 {
			t.Errorf("%svalue: expected %g, got %g", path, f, actual.Float)
		}

	case "bool":
		var b bool
		if err := json.Unmarshal(expected.Data, &b); err != nil {
			t.Errorf("%scannot parse expected bool data: %v", path, err)
			return
		}
		if actual.Bool != b {
			t.Errorf("%svalue: expected %v, got %v", path, b, actual.Bool)
		}

	case "null":
		// Nothing to compare

	case "array":
		if actual.Array == nil {
			t.Errorf("%sexpected array data, got nil", path)
			return
		}
		// Check element type
		var expectedElemType ValueType
		switch expected.ElementType {
		case "string":
			expectedElemType = TypeString
		case "int":
			expectedElemType = TypeInt
		case "float":
			expectedElemType = TypeFloat
		case "bool":
			expectedElemType = TypeBool
		case "array":
			expectedElemType = TypeArray
		case "null":
			expectedElemType = TypeNull
		}
		if actual.Array.ElementType != expectedElemType {
			t.Errorf("%selement_type: expected %v, got %v", path, expectedElemType, actual.Array.ElementType)
		}

		var items []expectedValue
		if err := json.Unmarshal(expected.Data, &items); err != nil {
			t.Errorf("%scannot parse expected array data: %v", path, err)
			return
		}
		if len(items) != len(actual.Array.Items) {
			t.Errorf("%sarray length: expected %d, got %d", path, len(items), len(actual.Array.Items))
			return
		}
		for i, item := range items {
			compareValue(t, fmt.Sprintf("%s[%d].", path, i), item, actual.Array.Items[i])
		}
	}
}
